src/main/kotlin/com/zepben/evolve/cim/iec61970/base/core/IdentifiedObject.kt

```diff
@@ -62,25 +62,29 @@ abstract class IdentifiedObject(mRID: String = "") {
      */
     fun getName(type: String, name: String): Name? = _names?.getByTypeAndName(type, name)
 
-    fun addName(name: Name): IdentifiedObject {
-        require(name.identifiedObject === this) { "Attempting to add a Name to ${typeNameAndMRID()} that does not reference this identified object" }
+    fun getName(type: NameType, name: String): Name? = _names?.getByTypeAndName(type.name, name)
 
-        if (getName(name.type.name, name.name) != null)
-            return this
+    fun addName(type: NameType, name: String): IdentifiedObject {
 
-        _names = _names ?: mutableListOf()
-        _names!!.add(name)
+        _names = _names ?: mutableSetOf()
+        _names!!.add(type.getOrAddName(name, this))
 
         return this
     }
 
     fun removeName(name: Name?): Boolean {
         val ret = _names?.remove(name) == true
+        // Remove names from nameType
+        if (ret) name?.type?.removeName(name)
```

Comment:
question: You are removing the `Name` from the `NameType` when you are removing the `Name` from the `IdentifiedObject`, but what do you expect to happen if, as a user, you directly remove the `Name` from the `NameType` (there are at least two functions that you can use to do this)?

Would you expect it to also remove the `Name` from the `IdentifiedObject`?

Reply:
Done the change and changed the test to reflect the change.